src/helpers/Event.ts

```diff
@@ -2,28 +2,51 @@ import { WriteStream } from 'fs'
 import { ServerResponse } from 'http'
 import { PokeError, PokeSuccess } from '../interfaces/PokeResult'
 
+type CallbackEvent = 'data' | 'error' | 'response' | 'end'
+
+function isCallbackEvent(input:string): input is CallbackEvent {
+    return /^data|error|response|end$/.test(input)
+}
+
+type Stream = WriteStream | ServerResponse
+
+type EventCallbacksContainer<Result> = {
+    [e in CallbackEvent | 'stream']? :
+    e extends 'data'
+      ? (chunk:unknown) => void // FIXME what is the type of chuck should be string?
+      : e extends 'error'
+          ? (result:PokeError<Result>) => void
+            :e extends 'response'
+            ? (param?:PokeSuccess<Result>) => void
+                : e extends 'end'
+                    ? () => void
+                    : e extends 'stream'
+                        ? Stream
+                        : never
+}
+
 interface EventManager <Result>{
-    set: (eventName:string, callback:(param?:unknown) => void) => void,
-    response: (result:PokeSuccess<Result>) => void,
-    end:() => void,
-    error: (result:PokeError<Result>) => void,
-    data: (chunk:string|unknown) => void,
+    set: <Event extends CallbackEvent>(eventName: Event, callback: EventCallbacksContainer<Result>[Event]) => void,
+    response: EventCallbacksContainer<Result>['response'],
+    end: EventCallbacksContainer<Result>['end'],
+    error: EventCallbacksContainer<Result>['error'],
+    data: EventCallbacksContainer<Result>['data'],
     stream: {
-        set: (writableStream:WriteStream|ServerResponse) => void,
-        write: (chunk:string|unknown) => void,
+        set: (writableStream: Stream) => void,
+        write: (chunk:unknown) => void, // FIXME what is the type of chuck should be string?
         end: () => void,
     }
 }
 
-const Event = function <Result>(): EventManager<Result> {
+const initEventManager = function <Result>(): EventManager<Result> {
     // the place to stores those callbacks
-    const callbacks = {}
+    const callbacks: EventCallbacksContainer<Result> = {}
     // return append callback methods and event callback methods
     return {
         // set 
         set: (eventName, callback) => {
             // valid event name and callback is a function
-            if(/^data|error|response|end$/.test(eventName) && /^function$/.test(typeof callback)) {
```

Comment:
removed some runtime check if a callback is a function, seems too unnecessary, could revert if they are really needed